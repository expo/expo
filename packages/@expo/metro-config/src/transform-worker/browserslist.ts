/**
 * Copyright Â© 2025 650 Industries.
 */

const debug = require('debug')('expo:metro:browserslist') as typeof console.log;

const browserslistCache: Record<string, import('lightningcss').Targets> = {};

export async function getBrowserslistTargets(
  projectRoot: string
): Promise<import('lightningcss').Targets> {
  if (browserslistCache[projectRoot]) {
    return browserslistCache[projectRoot];
  }
  const browserslist = await import('browserslist');
  const { browserslistToTargets } = await import('lightningcss');

  const targets = browserslistToTargets(
    browserslist.default(undefined, {
      throwOnMissing: false,
      ignoreUnknownVersions: true,
      path: projectRoot,
    })
  );

  debug('Browserslist targets: %O', targets);
  browserslistCache[projectRoot] = targets;
  return targets;
}
