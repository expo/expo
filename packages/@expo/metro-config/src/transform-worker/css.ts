export function pathToHtmlSafeName(path: string) {
  return path.replace(/[^a-zA-Z0-9_]/g, '_');
}

export function getHotReplaceTemplate(id: string) {
  // In dev mode, we need to replace the style tag instead of appending it
  // use the path as the expo-css-hmr attribute to find the style tag
  // to replace.
  const attr = JSON.stringify(pathToHtmlSafeName(id));
  return `style.setAttribute('data-expo-css-hmr', ${attr});
  const previousStyle = document.querySelector('[data-expo-css-hmr=${attr}]');
  if (previousStyle) {
    previousStyle.parentNode.removeChild(previousStyle);
  }`;
}

export function wrapDevelopmentCSS(props: { src: string; filename: string }) {
  const withBackTicksEscaped = escapeBackticksAndOctals(props.src);
  return `(() => {
  if (typeof document === 'undefined') {
    return
  }
  const head = document.head || document.getElementsByTagName('head')[0];
  const style = document.createElement('style');
  ${getHotReplaceTemplate(props.filename)}
  style.setAttribute('data-expo-loader', 'css');
  head.appendChild(style);
  const css = \`${withBackTicksEscaped}\`;
  if (style.styleSheet){
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
})();`;
}

export function escapeBackticksAndOctals(str: string) {
  if (typeof str !== 'string') {
    return '';
  }

  return str
    .replace(/\\/g, '\\\\')
    .replace(/`/g, '\\`')
    .replace(/[\0-\7]/g, (match) => `\\0${match.charCodeAt(0).toString(8)}`);
}
