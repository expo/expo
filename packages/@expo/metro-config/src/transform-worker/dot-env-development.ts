/**
 * Copyright 2025-present 650 Industries (Expo). All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as dotenv from 'dotenv';
import { expand as dotenvExpand } from 'dotenv-expand';

export function parseEnvFile(src: string, isClient: boolean): Record<string, string> {
  const expandedEnv: Record<string, string> = {};
  const envFileParsed = dotenv.parse(src);

  if (envFileParsed) {
    const allExpandedEnv = dotenvExpand({
      parsed: envFileParsed,
      processEnv: {},
    });

    for (const key of Object.keys(envFileParsed)) {
      if (allExpandedEnv.parsed?.[key]) {
        if (isClient && !key.startsWith('EXPO_PUBLIC_')) {
          // Don't include non-public variables in the client bundle.
          continue;
        }
        expandedEnv[key] = allExpandedEnv.parsed[key];
      }
    }
  }
  return expandedEnv;
}
