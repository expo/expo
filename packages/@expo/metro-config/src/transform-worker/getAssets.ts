/**
 * Copyright 2023-present 650 Industries (Expo). All rights reserved.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { AssetData, Module } from 'metro';
import { getAssetData } from 'metro/src/Assets';
import { getJsOutput, isJsModule } from 'metro/src/DeltaBundler/Serializers/helpers/js';
import assert from 'node:assert';
import crypto from 'node:crypto';
import path from 'node:path';

import { ReadOnlyDependencies } from '../serializer/getCssDeps';

type Options = {
  processModuleFilter: (modules: Module) => boolean;
  assetPlugins: readonly string[];
  platform?: string | null;
  projectRoot: string;
  publicPath: string;
};

function md5Hash(data: string[]) {
  if (data.length === 1) return data[0];
  const hash = crypto.createHash('md5');
  hash.update(data.join(''));
  return hash.digest('hex');
}

function assertHashedAssetData(data: AssetData): asserts data is HashedAssetData {
  assert(
    'fileHashes' in data,
    'Assets must have hashed files. Ensure the expo-asset plugin is installed.'
  );
}

export async function getUniversalAssetData(
  assetPath: string,
  localPath: string,
  assetDataPlugins: readonly string[],
  platform: string | null | undefined,
  publicPath: string
): Promise<HashedAssetData> {
  const data = await getAssetData(assetPath, localPath, assetDataPlugins, platform, publicPath);
  assertHashedAssetData(data);

  // NOTE(EvanBacon): This is where we modify the asset to include a hash in the name for web cache invalidation.
  if (platform === 'web' && publicPath.includes('?export_path=')) {
    // `local-image.[contenthash]`. Using `.` but this won't work if we ever apply to Android because Android res files cannot contain `.`.
    // TODO: Prevent one multi-res image from updating the hash in all images.
    // @ts-expect-error: name is typed as readonly.
    data.name = `${data.name}.${md5Hash(data.fileHashes)}`;
  }

  return data;
}

export type HashedAssetData = AssetData & { fileHashes: string[]; _name?: string };

export default async function getAssets(
  dependencies: ReadOnlyDependencies,
  options: Options
): Promise<HashedAssetData[]> {
  const promises: Promise<HashedAssetData>[] = [];
  const { processModuleFilter } = options;

  for (const module of dependencies.values()) {
    if (
      isJsModule(module) &&
      processModuleFilter(module) &&
      getJsOutput(module).type === 'js/module/asset' &&
      path.relative(options.projectRoot, module.path) !== 'package.json'
    ) {
      promises.push(
        getUniversalAssetData(
          module.path,
          path.relative(options.projectRoot, module.path),
          options.assetPlugins,
          options.platform,
          options.publicPath
        )
      );
    }
  }

  return await Promise.all(promises);
}
