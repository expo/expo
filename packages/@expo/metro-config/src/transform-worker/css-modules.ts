import { wrapDevelopmentCSS } from './css';

const RNW_CSS_CLASS_ID = '_';

export async function transformCssModuleWeb(props: {
  filename: string;
  src: string;
  options: { projectRoot: string; minify: boolean; dev: boolean; sourceMap: boolean };
}) {
  const { transform } = require('lightningcss') as typeof import('lightningcss');

  // TODO: Add bundling to resolve imports
  // https://lightningcss.dev/bundling.html#bundling-order

  const cssResults = transform({
    filename: props.filename,
    code: Buffer.from(props.src),
    sourceMap: props.options.sourceMap,
    cssModules: {
      // Prevent renaming CSS variables to ensure
      // variables created in global files are available.
      dashedIdents: false,
    },
    // cssModules: true,
    projectRoot: props.options.projectRoot,
    minify: props.options.minify,
  });
  const codeAsString = cssResults.code.toString();

  const { styles, reactNativeWeb, variables } = convertLightningCssToReactNativeWebStyleSheet(
    cssResults.exports!
  );

  let outputModule = `module.exports=Object.assign(${JSON.stringify(
    styles
  )},{unstable_styles:${JSON.stringify(reactNativeWeb)}},${JSON.stringify(variables)});`;

  if (props.options.dev) {
    const runtimeCss = wrapDevelopmentCSS({
      ...props,
      src: codeAsString,
    });

    outputModule += '\n' + runtimeCss;
  }

  return {
    output: outputModule,
    css: cssResults.code,
    map: cssResults.map,
  };
}

export function convertLightningCssToReactNativeWebStyleSheet(
  input: import('lightningcss').CSSModuleExports
) {
  const styles: Record<string, string> = {};
  const reactNativeWeb: Record<string, any> = {};
  const variables: Record<string, string> = {};
  // e.g. { container: { name: 'ahs8IW_container', composes: [], isReferenced: false }, }
  Object.entries(input).map(([key, value]) => {
    // order matters here
    let className = value.name;

    if (value.composes.length) {
      className += ' ' + value.composes.map((value) => value.name).join(' ');
    }

    // CSS Variables will be `{string: string}`
    if (key.startsWith('--')) {
      variables[key] = className;
    }

    styles[key] = className;
    reactNativeWeb[key] = { $$css: true, [RNW_CSS_CLASS_ID]: className };
    return {
      [key]: { $$css: true, [RNW_CSS_CLASS_ID]: className },
    };
  });

  return { styles, reactNativeWeb, variables };
}

export function matchCssModule(filePath: string): boolean {
  return !!/\.module(\.(native|ios|android|web))?\.(css|s[ac]ss)$/.test(filePath);
}
