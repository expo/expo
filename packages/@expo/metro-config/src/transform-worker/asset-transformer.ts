/**
 * Copyright 2023-present 650 Industries (Expo). All rights reserved.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Fork of the upstream transformer, but with modifications made for web production hashing.
 * https://github.com/facebook/metro/blob/412771475c540b6f85d75d9dcd5a39a6e0753582/packages/metro-transform-worker/src/utils/assetTransformer.js#L1
 */
import { generateAssetCodeFileAst } from 'metro/src/Bundler/util';
import { BabelTransformerArgs } from 'metro-babel-transformer';
import path from 'node:path';

import { getUniversalAssetData } from './getAssets';

export async function transform(
  { filename, options }: BabelTransformerArgs,
  assetRegistryPath: string,
  assetDataPlugins: readonly string[]
) {
  options ??= options || {
    platform: '',
    projectRoot: '',
    inlineRequires: false,
    minify: false,
  };

  const absolutePath = path.resolve(options.projectRoot, filename);

  const data = await getUniversalAssetData(
    absolutePath,
    filename,
    assetDataPlugins,
    options.platform,
    options.publicPath
  );

  return {
    ast: generateAssetCodeFileAst(assetRegistryPath, data),
  };
}
