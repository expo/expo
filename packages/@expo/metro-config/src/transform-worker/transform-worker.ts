/**
 * Copyright 2023-present 650 Industries (Expo). All rights reserved.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import countLines from 'metro/src/lib/countLines';
import type {
  JsTransformerConfig,
  JsTransformOptions,
  TransformResponse,
} from 'metro-transform-worker';

import { wrapDevelopmentCSS } from './css';
import { matchCssModule, transformCssModuleWeb } from './css-modules';
import * as worker from './metro-transform-worker';
import { transformPostCssModule } from './postcss';
import { compileSass, matchSass } from './sass';
import { ExpoJsOutput, JsOutput } from '../serializer/jsOutput';

export async function transform(
  config: JsTransformerConfig,
  projectRoot: string,
  filename: string,
  data: Buffer,
  options: JsTransformOptions
): Promise<TransformResponse> {
  const isCss = options.type !== 'asset' && /\.(s?css|sass)$/.test(filename);
  // If the file is not CSS, then use the default behavior.
  if (!isCss) {
    const environment = options.customTransformOptions?.environment;
    const isClientEnvironment = environment !== 'node' && environment !== 'react-server';
    if (
      isClientEnvironment &&
      // TODO: Ensure this works with windows.
      (filename.match(new RegExp(`^app/\\+html(\\.${options.platform})?\\.([tj]sx?|[cm]js)?$`)) ||
        // Strip +api files.
        filename.match(/\+api(\.(native|ios|android|web))?\.[tj]sx?$/))
    ) {
      // Remove the server-only +html file and API Routes from the bundle when bundling for a client environment.
      return worker.transform(
        config,
        projectRoot,
        filename,
        !options.minify
          ? Buffer.from(
              // Use a string so this notice is visible in the bundle if the user is
              // looking for it.
              '"> The server-only file was removed from the client JS bundle by Expo CLI."'
            )
          : Buffer.from(''),
        options
      );
    }

    if (
      isClientEnvironment &&
      !filename.match(/\/node_modules\//) &&
      filename.match(/\+api(\.(native|ios|android|web))?\.[tj]sx?$/)
    ) {
      // Clear the contents of +api files when bundling for the client.
      // This ensures that the client doesn't accidentally use the server-only +api files.
      return worker.transform(config, projectRoot, filename, Buffer.from(''), options);
    }

    return worker.transform(config, projectRoot, filename, data, options);
  }

  // If the platform is not web, then return an empty module.
  if (options.platform !== 'web') {
    const code = matchCssModule(filename) ? 'module.exports={ unstable_styles: {} };' : '';
    return worker.transform(
      config,
      projectRoot,
      filename,
      // TODO: Native CSS Modules
      Buffer.from(code),
      options
    );
  }

  let code = data.toString('utf8');

  // Apply postcss transforms
  const postcssResults = await transformPostCssModule(projectRoot, {
    src: code,
    filename,
  });

  if (postcssResults.hasPostcss) {
    code = postcssResults.src;
  }

  // TODO: When native has CSS support, this will need to move higher up.
  const syntax = matchSass(filename);
  if (syntax) {
    code = compileSass(projectRoot, { filename, src: code }, { syntax }).src;
  }

  // If the file is a CSS Module, then transform it to a JS module
  // in development and a static CSS file in production.
  if (matchCssModule(filename)) {
    const results = await transformCssModuleWeb({
      filename,
      src: code,
      options: {
        projectRoot,
        dev: options.dev,
        minify: options.minify,
        sourceMap: false,
      },
    });

    const jsModuleResults = await worker.transform(
      config,
      projectRoot,
      filename,
      Buffer.from(results.output),
      options
    );

    const cssCode = results.css.toString();
    const output: JsOutput[] = [
      {
        type: 'js/module',
        data: {
          ...jsModuleResults.output[0]?.data,

          // Append additional css metadata for static extraction.
          css: {
            code: cssCode,
            lineCount: countLines(cssCode),
            map: [],
            functionMap: null,
          },
        },
      },
    ];

    return {
      dependencies: jsModuleResults.dependencies,
      output,
    };
  }

  // Global CSS:

  const { transform } = require('lightningcss') as typeof import('lightningcss');

  // TODO: Add bundling to resolve imports
  // https://lightningcss.dev/bundling.html#bundling-order

  const cssResults = transform({
    filename,
    code: Buffer.from(code),
    sourceMap: false,
    cssModules: false,
    projectRoot,
    minify: options.minify,
  });

  // TODO: Warnings:
  // cssResults.warnings.forEach((warning) => {
  // });

  // Create a mock JS module that exports an empty object,
  // this ensures Metro dependency graph is correct.
  const jsModuleResults = await worker.transform(
    config,
    projectRoot,
    filename,
    options.dev ? Buffer.from(wrapDevelopmentCSS({ src: code, filename })) : Buffer.from(''),
    options
  );

  const cssCode = cssResults.code.toString();

  // In production, we export the CSS as a string and use a special type to prevent
  // it from being included in the JS bundle. We'll extract the CSS like an asset later
  // and append it to the HTML bundle.
  const output: ExpoJsOutput[] = [
    {
      type: 'js/module',
      data: {
        ...(jsModuleResults.output[0] as ExpoJsOutput).data,

        // Append additional css metadata for static extraction.
        css: {
          code: cssCode,
          lineCount: countLines(cssCode),
          map: [],
          functionMap: null,
          // Disable caching for CSS files when postcss is enabled and has been run on the file.
          // This ensures that things like tailwind can update on every change.
          skipCache: postcssResults.hasPostcss,
        },
      },
    },
  ];

  return {
    dependencies: jsModuleResults.dependencies,
    output,
  };
}

/**
 * A custom Metro transformer that adds support for processing Expo-specific bundler features.
 * - Global CSS files on web.
 * - CSS Modules on web.
 * - TODO: Tailwind CSS on web.
 */
module.exports = {
  // Use defaults for everything that's not custom.
  ...worker,
  transform,
};
