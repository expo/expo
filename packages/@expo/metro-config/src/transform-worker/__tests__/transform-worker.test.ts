import upstreamTransformer, { JsTransformOptions } from 'metro-transform-worker';

import { transform } from '../transform-worker';

jest.mock('metro-transform-worker', () => ({
  transform: jest.fn(),
}));

beforeEach(() => {
  jest.mocked(upstreamTransformer.transform).mockReset();
});

const doTransformForOutput = async (
  filename: string,
  src: string,
  options: Partial<JsTransformOptions>
): Promise<{ input: string; output: any }> => {
  jest.mocked(upstreamTransformer.transform).mockResolvedValueOnce({
    dependencies: [],
    output: [
      {
        data: {},
      },
    ],
  });
  const output = await doTransform(filename, src, options);
  expect(upstreamTransformer.transform).toBeCalledTimes(1);
  return {
    input: jest.mocked(upstreamTransformer.transform).mock.calls[0][3].toString('utf8'),
    output,
  };
};

const doTransformForInput = async (
  filename: string,
  src: string,
  options: Partial<JsTransformOptions>
): Promise<string> => {
  await doTransform(filename, src, options);
  expect(upstreamTransformer.transform).toBeCalledTimes(1);
  return jest.mocked(upstreamTransformer.transform).mock.calls[0][3].toString('utf8');
};
const doTransform = async (filename: string, src: string, options: Partial<JsTransformOptions>) => {
  return transform({} as any, '/', filename, Buffer.from(src), {
    dev: true,
    minify: false,
    platform: 'web',
    type: 'script',
    hot: false,
    inlinePlatform: false,
    inlineRequires: false,
    unstable_transformProfile: 'default',
    ...options,
  });
};

it(`performs a sanity check by transforming a JS file as expected`, async () => {
  expect(
    await doTransformForInput('acme.js', 'export default {}', {
      dev: true,
      minify: false,
      platform: 'web',
    })
  ).toMatchInlineSnapshot(`"export default {}"`);
});

it(`transforms a global CSS file in dev for web`, async () => {
  expect(
    await doTransformForOutput('acme.css', 'body { background: red; }', {
      dev: true,
      minify: false,
      platform: 'web',
    })
  ).toMatchSnapshot();
});

it(`transforms a global CSS file in dev for native`, async () => {
  expect(
    await doTransformForInput('acme.css', 'body { background: red; }', {
      dev: true,
      minify: false,
      platform: 'ios',
    })
  ).toMatchInlineSnapshot(`""`);
});

describe('CSS Modules', () => {
  describe('ios', () => {
    it(`transforms for dev, minified`, async () => {
      expect(
        await doTransformForInput('acme.module.css', '.container { background: red; }', {
          dev: true,
          minify: true,
          platform: 'ios',
        })
      ).toMatchInlineSnapshot(`"module.exports={ unstable_styles: {} };"`);
    });
    it(`transforms for dev, not minified`, async () => {
      expect(
        await doTransformForInput('acme.module.css', '.container { background: red; }', {
          dev: true,
          minify: false,
          platform: 'ios',
        })
      ).toMatchInlineSnapshot(`"module.exports={ unstable_styles: {} };"`);
    });

    it(`transforms for prod, minified`, async () => {
      expect(
        await doTransformForInput('acme.module.css', '.container { background: red; }', {
          dev: false,
          minify: true,
          platform: 'ios',
        })
      ).toMatchInlineSnapshot(`"module.exports={ unstable_styles: {} };"`);
    });

    it(`transforms for prod, not minified`, async () => {
      expect(
        await doTransformForInput('acme.module.css', '.container { background: red; }', {
          dev: false,
          minify: false,
          platform: 'ios',
        })
      ).toMatchInlineSnapshot(`"module.exports={ unstable_styles: {} };"`);
    });
  });
  describe('web', () => {
    it(`transforms for dev, minified`, async () => {
      expect(
        await doTransformForOutput('acme.module.css', '.container { background: red; }', {
          dev: true,
          minify: true,
          platform: 'web',
        })
      ).toMatchSnapshot();
    });
    it(`transforms for dev, not minified`, async () => {
      expect(
        await doTransformForOutput('acme.module.css', '.container { background: red; }', {
          dev: true,
          minify: false,
          platform: 'web',
        })
      ).toMatchSnapshot();
    });

    it(`transforms for prod, minified`, async () => {
      const { input, output } = await doTransformForOutput(
        'acme.module.css',
        '.container { background: red; }',
        {
          dev: false,
          minify: true,
          platform: 'web',
        }
      );
      expect(input).toMatchSnapshot();

      // no HMR code
      expect(input).not.toMatch(/document\./);
      // Has keys
      expect(input).toMatch(/_R_BGG_container/);
      expect(output).toEqual({
        dependencies: [],
        output: [
          {
            data: {
              // Required CSS metadata for static export
              css: {
                code: '._R_BGG_container{background:red}',
                functionMap: null,
                lineCount: 1,
                map: [],
              },
            },
            type: 'js/module',
          },
        ],
      });
    });

    it(`transforms for prod, not minified`, async () => {
      const { input, output } = await doTransformForOutput(
        'acme.module.css',
        '.container { background: red; }',
        {
          dev: false,
          minify: false,
          platform: 'web',
        }
      );
      expect(input).toMatchSnapshot();

      // no HMR code
      expect(input).not.toMatch(/document\./);
      // Has keys
      expect(input).toMatch(/_R_BGG_container/);
      expect(output).toEqual({
        dependencies: [],
        output: [
          {
            data: {
              // Required CSS metadata for static export
              css: {
                // Not minified, formatted actually.
                code: ['._R_BGG_container {', '  background: red;', '}', ''].join('\n'),
                functionMap: null,
                lineCount: 4,
                map: [],
              },
            },
            type: 'js/module',
          },
        ],
      });
    });
  });
});

// TODO: Test +api files to ensure all extensions work
describe('Expo Router server files (+html, +api)', () => {
  const matchable = /> The server-only file was removed from the client JS bundle by Expo CLI/;
  it(`strips +html file from client bundles`, async () => {
    for (const file of [
      'app/+html.js',
      'app/+html.ts',
      'app/+html.tsx',
      'app/+html.web.jsx',
      'app/+html.web.ts',
    ]) {
      jest.mocked(upstreamTransformer.transform).mockReset();

      expect(
        (
          await doTransformForOutput(file, 'REMOVE ME!', {
            dev: true,
            minify: false,
            customTransformOptions: {
              environment: 'client',
            },
            platform: 'web',
          })
        ).input
      ).toMatch(matchable);
    }

    // Ensure the server code doesn't leak into the client on any platform.
    for (const platform of ['ios', 'android', 'web']) {
      jest.mocked(upstreamTransformer.transform).mockReset();
      expect(
        (
          await doTransformForOutput('app/+html.js', 'REMOVE ME!', {
            dev: true,
            minify: false,
            customTransformOptions: {
              environment: 'client',
            },
            platform,
          })
        ).input
      ).toMatch(matchable);
    }
  });
  it(`strips without warning when minify is enabled`, async () => {
    expect(
      (
        await doTransformForOutput('app/+html.js', 'KEEP', {
          dev: false,
          minify: true,
          customTransformOptions: {
            environment: 'client',
          },
          platform: 'web',
        })
      ).input
    ).toMatch('');
  });
  it(`modifies server files even if no server indication is provided`, async () => {
    expect(
      (
        await doTransformForOutput('app/+html.js', 'KEEP', {
          dev: true,
          minify: false,
          platform: 'web',
        })
      ).input
    ).toMatch(matchable);
  });
  it(`preserves when bundling for Node.js environments`, async () => {
    expect(
      (
        await doTransformForOutput('app/+html.js', 'KEEP', {
          dev: true,
          minify: false,
          customTransformOptions: {
            environment: 'node',
          },
          platform: 'ios',
        })
      ).input
    ).toMatch('KEEP');
  });
});
