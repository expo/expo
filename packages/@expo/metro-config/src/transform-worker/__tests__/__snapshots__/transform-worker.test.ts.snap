// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSS Modules web transforms for dev, minified 1`] = `
{
  "input": "module.exports=Object.assign({"container":"_R_BGG_container"},{unstable_styles:{"container":{"$$css":true,"_":"_R_BGG_container"}}}{});
(() => {
  if (typeof document === 'undefined') {
    return
  }
  const head = document.head || document.getElementsByTagName('head')[0];
  const style = document.createElement('style');
  style.setAttribute('data-expo-css-hmr', "acme_module_css");
  const previousStyle = document.querySelector('[data-expo-css-hmr="acme_module_css"]');
  if (previousStyle) {
    previousStyle.parentNode.removeChild(previousStyle);
  }
  style.setAttribute('data-expo-loader', 'css');
  head.appendChild(style);
  const css = \`._R_BGG_container{background:red}\`;
  if (style.styleSheet){
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
})();",
  "output": {
    "dependencies": [],
    "output": [
      {
        "data": {
          "css": {
            "code": "._R_BGG_container{background:red}",
            "functionMap": null,
            "lineCount": 1,
            "map": [],
          },
        },
        "type": "js/module",
      },
    ],
  },
}
`;

exports[`CSS Modules web transforms for dev, not minified 1`] = `
{
  "input": "module.exports=Object.assign({"container":"_R_BGG_container"},{unstable_styles:{"container":{"$$css":true,"_":"_R_BGG_container"}}}{});
(() => {
  if (typeof document === 'undefined') {
    return
  }
  const head = document.head || document.getElementsByTagName('head')[0];
  const style = document.createElement('style');
  style.setAttribute('data-expo-css-hmr', "acme_module_css");
  const previousStyle = document.querySelector('[data-expo-css-hmr="acme_module_css"]');
  if (previousStyle) {
    previousStyle.parentNode.removeChild(previousStyle);
  }
  style.setAttribute('data-expo-loader', 'css');
  head.appendChild(style);
  const css = \`._R_BGG_container {
  background: red;
}
\`;
  if (style.styleSheet){
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
})();",
  "output": {
    "dependencies": [],
    "output": [
      {
        "data": {
          "css": {
            "code": "._R_BGG_container {
  background: red;
}
",
            "functionMap": null,
            "lineCount": 4,
            "map": [],
          },
        },
        "type": "js/module",
      },
    ],
  },
}
`;

exports[`CSS Modules web transforms for prod, minified 1`] = `"module.exports=Object.assign({"container":"_R_BGG_container"},{unstable_styles:{"container":{"$$css":true,"_":"_R_BGG_container"}}}{});"`;

exports[`CSS Modules web transforms for prod, not minified 1`] = `"module.exports=Object.assign({"container":"_R_BGG_container"},{unstable_styles:{"container":{"$$css":true,"_":"_R_BGG_container"}}}{});"`;

exports[`transforms a global CSS file in dev for web 1`] = `
{
  "input": "(() => {
  if (typeof document === 'undefined') {
    return
  }
  const head = document.head || document.getElementsByTagName('head')[0];
  const style = document.createElement('style');
  style.setAttribute('data-expo-css-hmr', "acme_css");
  const previousStyle = document.querySelector('[data-expo-css-hmr="acme_css"]');
  if (previousStyle) {
    previousStyle.parentNode.removeChild(previousStyle);
  }
  style.setAttribute('data-expo-loader', 'css');
  head.appendChild(style);
  const css = \`body { background: red; }\`;
  if (style.styleSheet){
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
})();",
  "output": {
    "dependencies": [],
    "output": [
      {
        "data": {
          "css": {
            "code": "body {
  background: red;
}
",
            "functionMap": null,
            "lineCount": 4,
            "map": [],
          },
        },
        "type": "js/module",
      },
    ],
  },
}
`;
